src/debug/gdbstub.rs

```diff
@@ -446,20 +445,16 @@ pub mod svsm_gdbstub {
                     let context = (self.ctx as *const TaskContext).as_ref().unwrap();
                     *regs = X86_64CoreRegs::from(context);
                 }
+            } else if let Some(task) = TASKLIST.lock().get_task(tid.get() as u32) {
+                // The registers are stored in the top of the task stack as part of the
+                // saved context. We need to switch to the task pagetable to access them.
+                let _task_context = GdbTaskContext::switch_to_task(tid.get() as u32);
+                unsafe {
+                    *regs = X86_64CoreRegs::from(&*(task.rsp as *const TaskContext));
+                };
+                regs.regs[7] = task.rsp;
```

Comment:
The previous code deliberately assigned the task `Option` to a variable before checking if it was `Some` or `None` as otherwise the `TASKLIST` lock is held until the if scope is exited, causing a deadlock in `switch_to_task()`. This is what was causing the thread list to deadlock as GDB queries the register state of each task in turn.

```suggestion
            } else {
                let task = TASKLIST.lock().get_task(tid.get() as u32);
                if let Some(task) = task {
                    // The registers are stored in the top of the task stack as part of the
                    // saved context. We need to switch to the task pagetable to access them.
                    let _task_context = GdbTaskContext::switch_to_task(tid.get() as u32);
                    unsafe {
                        *regs = X86_64CoreRegs::from(&*(task.rsp as *const TaskContext));
                    };
                    regs.regs[7] = task.rsp;
```

Reply:
Thanks for fixing this up. I merged this change into the patch `SVSM/task: Move task scheduling state into separate struct`